File: Client/Assets/uLua/Lua/MyLua/Libs/Network/Cmd/NetCmdHandleBase.lua
local M = GlobalNS.Class(GlobalNS.GObject);
M.clsName = "NetCmdHandleBase";
GlobalNS[M.clsName] = M;

function M:ctor()
    self.m_id2HandleDic = GlobalNS.new(GlobalNS.MDictionary);
end

function M:addParamHandle(paramId, pThis, func)
    if(not self.m_id2HandleDic:ContainsKey(paramId)) then
        local disp = GlobalNS.new(GlobalNS.AddOnceEventDispatch);
        self.m_id2HandleDic:Add(paramId, disp);
    else
        GCtx.mLogSys.log("Msg Id Already Register", GlobalNS.LogTypeId.eLogCommon);
    end

    self.m_id2HandleDic:value(paramId):addEventHandle(pThis, func);
end

function M:removeParamHandle(paramId, pThis, func)
    if(self.m_id2HandleDic.ContainsKey(paramId)) then
        self.m_id2HandleDic:value(paramId):removeEventHandle(pThis, func);
    else
        GCtx.mLogSys.log("ParamId not Register", GlobalNS.LogTypeId.eLogCommon);
    end
end

function M:call(dispObj)
    local cmd = dispObj;
    self:handleMsg(cmd.bu, cmd.byCmd, cmd.byParam);
end

function M:handleMsg(bu, byCmd, byParam)
    GCtx.mLogSys:log("NetCmdHandleBase Start handleMsg", GlobalNS.LogTypeId.eLogCommon);
    if(self.m_id2HandleDic:ContainsKey(byParam)) then
        GCtx.mLogSys:log("NetCmdHandleBase In handleMsg", GlobalNS.LogTypeId.eLogCommon);
        self.m_id2HandleDic:value(byParam):dispatchEvent(bu);
    else
        
    end
end

return M;

File: Client/Assets/Scripts/Libs/Network/Cmd/NetCmdHandleBase.cs
using System;
using System.Collections.Generic;

namespace SDK.Lib
{
    public class NetCmdHandleBase : ICalleeObject
    {
        protected Dictionary<int, AddOnceEventDispatch> m_id2HandleDic;

        public NetCmdHandleBase()
        {
            m_id2HandleDic = new Dictionary<int, AddOnceEventDispatch>();
        }

        public void addParamHandle(int paramId, Action<IDispatchObject> handle)
        {
            if(!m_id2HandleDic.ContainsKey(paramId))
            {
                m_id2HandleDic[paramId] = new AddOnceEventDispatch();   
            }
            else
            {
                Ctx.m_instance.m_logSys.log("Msg Id Already Register");
            }

            m_id2HandleDic[paramId].addEventHandle(null, handle);
        }

        public void removeParamHandle(int paramId, Action<IDispatchObject> handle)
        {
            if(m_id2HandleDic.ContainsKey(paramId))
            {
                m_id2HandleDic[paramId].removeEventHandle(null, handle);
            }
            else
            {
                Ctx.m_instance.m_logSys.log("ParamId not Register");
            }
        }

        public void call(IDispatchObject dispObj)
        {

        }

        public virtual void handleMsg(ByteBuffer bu, byte byCmd, byte byParam)
        {
            if(m_id2HandleDic.ContainsKey(byParam))
            {
                m_id2HandleDic[byParam].dispatchEvent(bu);
            }
            else
            {
                Ctx.m_instance.m_logSys.log(string.Format("消息没有处理: byCmd = {0},  byParam = {1}", byCmd, byParam));
            }
        }
    }
}